lib/src/embedded/dispatcher.ts

```diff
@@ -0,0 +1,286 @@
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import {Observable, Subject, Subscription} from 'rxjs';
+import {filter, map} from 'rxjs/operators';
+
+import {InboundMessage, OutboundMessage} from '../vendor/embedded_sass_pb';
+import {
+  InboundRequest,
+  InboundRequestType,
+  InboundResponse,
+  InboundResponseType,
+  InboundTypedMessage,
+  OutboundRequest,
+  OutboundRequestType,
+  OutboundResponse,
+  OutboundResponseType,
+  OutboundTypedMessage,
+} from './message-transformer';
+import {RequestTracker} from './request-tracker';
+
+type PromiseOr<T> = T | Promise<T>;
+
+/**
+ * Dispatches requests, responses, and events.
+ *
+ * Consumers can send an inbound request. This returns a promise that resolves
+ * with the corresponding outbound response, or errors if any Protocol Errors
+ * were encountered.
+ *
+ * Consumers can also register callbacks for processing different types of
+ * outbound requests. When an outbound request arrives, this runs the
+ * appropriate callback to process it, then sends the result inbound. Errors
+ * are not thrown to outbound request listeners; outbound request streams die
+ * quietly.
+ *
+ * Requests and responses do not need to have their ID set, since the dispatcher
+ * handles setting the ID on all messages.
+ *
+ * Detects ProtocolErrors caused by mismatched requests and responses.
+ */
+export class Dispatcher {
+  // Tracks the IDs of all inbound requests. An outbound response with matching
+  // ID and type will remove the ID.
+  private readonly pendingInboundRequests = new RequestTracker();
+
+  // Tracks the IDs of all outbound requests. An inbound response with matching
+  // ID and type will remove the ID.
+  private readonly pendingOutboundRequests = new RequestTracker();
+
+  // All outbound messages. If we detect any errors while dispatching messages,
+  // this completes.
+  private readonly messages$ = new Subject<OutboundTypedMessage>();
+
+  // Collects all errors that we might encounter. An emitted error shuts down
+  // all promises awaiting an outbound response.
+  private readonly error$ = new Subject<Error>();
+
+  /** All outbound log events. */
+  readonly logEvents$ = this.messages$.pipe(
+    filter(message => message.type === OutboundMessage.MessageCase.LOGEVENT),
+    map(message => message.payload as OutboundMessage.LogEvent)
+  );
+
+  constructor(
+    private readonly outboundMessages$: Observable<OutboundTypedMessage>,
+    private readonly writeInboundMessage: (message: InboundTypedMessage) => void
+  ) {
+    this.outboundMessages$.subscribe(
+      message => {
+        try {
+          this.registerOutboundMessage(message);
+          this.messages$.next(message);
+        } catch (error) {
+          this.error$.next(error);
+        }
+      },
+      error => this.error$.next(error),
+      () => {
+        this.messages$.complete();
+        this.error$.complete();
+      }
+    );
```

Comment:
Consider rewriting as:

```suggestion
    this.outboundMessages$.pipe(
      tap(message => this.registerOutboundMessage(message)),
    ).subscribe(
      (message) => this.messages$.next(message),
      (error) => this.error$.error(error),
      () => {
        this.messages$.complete();
        this.error$.complete();
      }
    });
```

This more functional style avoids a manual `try`/`catch` around `registerOutboundMessage()` since any errors there are folded back into the pipeline.